src/storage/mvcc/reader/backward_scanner.rs

```diff
@@ -0,0 +1,713 @@
+                    assert!(self.lock_cursor.valid());
+                    let lock_value = self.lock_cursor.value(&mut self.statistics.lock);
+                    let lock = Lock::parse(lock_value)?;
+                    safe_ts = super::util::check_lock(user_key, safe_ts, &lock)?
+                }
+            }
+            IsolationLevel::RC => {}
+        }
+
+        if !has_write {
+            return Ok(None);
+        }
+        assert!(self.write_cursor.valid());
+
+        // At first, we try to use several `prev()` to get the desired version.
+
+        // We need to save last desired version, because when we may move to an unwanted version
+        // at any time.
+        let mut last_version = None;
+        let mut last_checked_commit_ts = 0;
+
+        for i in 0..REVERSE_SEEK_BOUND {
+            if i > 0 {
+                // We are already pointing at the smallest version, so we don't need to prev()
+                // for the first iteration. So we will totally call `prev()` function
+                // `REVERSE_SEEK_BOUND - 1` times.
+                self.write_cursor.prev(&mut self.statistics.write);
+                if !self.write_cursor.valid() {
+                    // Key space ended. We use `last_version` as the return.
+                    return Ok(self.handle_last_version(last_version, user_key)?);
+                }
+            }
+
+            let mut is_done = false;
+            {
+                let current_key = self.write_cursor.key(&mut self.statistics.write);
+                last_checked_commit_ts = Key::decode_ts_from(current_key)?;
+
+                if !Key::is_user_key_eq(current_key, user_key.encoded().as_slice())
+                    || last_checked_commit_ts > safe_ts
+                {
+                    // Meet another key or meet an unwanted version. We use `last_version` as the return.
+                    is_done = true;
+                }
+            }
+            if is_done {
+                return Ok(self.handle_last_version(last_version, user_key)?);
+            }
+
+            let write = Write::parse(self.write_cursor.value(&mut self.statistics.write))?;
+            self.statistics.write.processed += 1;
+
+            match write.write_type {
+                WriteType::Put | WriteType::Delete => last_version = Some(write),
+                WriteType::Lock | WriteType::Rollback => {}
+            }
+        }
+
+        // At this time, we must have current commit_ts <= safe_ts. If commit_ts == safe_ts,
+        // we don't need to seek any more and we can just utilize `last_version`.
+        if last_checked_commit_ts == safe_ts {
+            return Ok(self.handle_last_version(last_version, user_key)?);
+        }
+
+        // After several `prev()`, we still not get the latest version for the specified ts,
+        // use seek to locate the latest version.
+        let last_handled_key = self.write_cursor.key(&mut self.statistics.write).to_vec();
```

Comment:
Can we use `last_checked_commit_ts` instead of `last_handled_key`? Comparing two keys may be slower than decoding ts from slice and comparing two u64. You should bench it.

Reply:
good idea.

Reply:
@MyonKeminta